pkg/netutil/netutil.go

```diff
@@ -50,14 +52,23 @@ func NewCNIEnv(cniPath, cniConfPath string) (*CNIEnv, error) {
 	if err := os.MkdirAll(e.NetconfPath, 0755); err != nil {
 		return nil, err
 	}
-	if err := e.ensureDefaultNetworkConfig(); err != nil {
-		return nil, err
-	}
 	networks, err := e.networkConfigList()
 	if err != nil {
 		return nil, err
 	}
 	e.Networks = networks
+
+	// NOTE: must come after `e.networkConfigList()`.
+	defaultNet, err := e.getDefaultNetworkConfig()
+	if err != nil {
+		return nil, fmt.Errorf("failed to check for default network: %s", err)
+	}
+	if defaultNet == nil {
+		if err := e.createDefaultNetworkConfig(); err != nil {
+			return nil, fmt.Errorf("failed to create default network: %s", err)
+		}
+	}
```

Comment:
can we  move`ensureDefaultNetworkConfig` after  `e.networkConfigList()` and then put that in  ?

Reply:
I'm not sure what you mean since `ensureDefaultNetworkConfig()` has now become `createDefaultNetworkConfig()`. (which is indeed now called after `e.networkConfigList()`)

It has however just hit me that creating the default network after the `networkConfigList()` will mean that `e.Networks` will not include the newly-created default network.

While I could just go `e.Networks = append(e.Networks, newDefaultNetwork)`, I still think this shows a deeper underlying problem in how we expose this field:
- [`e.Networks` is a public slice of a private type (`networkConfig`)](https://github.com/containerd/nerdctl/blob/main/pkg/netutil/netutil.go#L42)
- `e.Networks` is [only set once, and that is in a constructor function in this same module](https://github.com/containerd/nerdctl/blob/main/pkg/netutil/netutil.go#L60), so there's no need for it to be public
- it's used in `cmd/nerdctl/nerwork_*.go` but only to be iterated over, it's not actually set anywhere outside this module
- `e.NetworkMap()` is just a convenience method which has the potential to "miss" networks if they share names by [overriding any names already in the map](https://github.com/containerd/nerdctl/blob/main/pkg/netutil/netutil.go#L67) from whatever it finds in `e.Networks`

I'm guessing `e.Networks` might have been added for caching reasons or something but at this stage IMO it just makes the network listing logic less consistent and should be removed entirely, what do you think?

Reply:
- SGTM regarding introducing `NetworkList()` to expose networkConfigs
- What I propose is to wrap the `e.getDefaultNetworkConfig()` and `createDefaultNetworkConfig()` into `ensureDefaultNetworkConfig()`

stgh like this 
```
func (e *CNIEnv) ensureDefaultNetworkConfig() error{
	defaultNet, err := e.GetDefaultNetworkConfig()
	if err != nil {
		return  fmt.Errorf("failed to check for default network: %s", err)
	}
	if defaultNet == nil {
		if err := e.createDefaultNetworkConfig(); err != nil {
			return fmt.Errorf("failed to create default network: %s", err)
		}
	}
        return nil
}
```